### ISSUE TITLE
Expired single-cell tombstones are kept in cache indefinitely

### ISSUE THREAD

**Issue by mykaul** [scylladb/scylladb#16093]
Follow-up to https://github.com/scylladb/scylladb/issues/6033 , to deal specifically with single cell tombstones.

**Comment by bhalevy** [comment]
@tgrabiec/@michoecho please shed more light on this issue.
What exactly needs to be fixed and what is the plan to fix it.

**Comment by michoecho** [comment]
The issue is as described by the title. There is no mechanism which expires cell tombstones out from the cache, hence they remain in the cache (and take up space, and CPU cycles on access) until the entire row is evicted. If a row remains in the cache indefinitely, so will the tombstones.

AFAIK there is currently no plan to fix this. But if we had to fix it, I guess the solution would be to expire cell tombstones during version merges (in calls to `mutation_partition_v2::apply_monotonically` from cache) and during reads (in `cache_flat_mutation_reader::add_clustering_row_to_buffer`).

Does this deserve P1? Intuitively, it seems like a relatively minor optimization.

**Comment by avikivity** [comment]
> The issue is as described by the title. There is no mechanism which expires cell tombstones out from the cache, hence they remain in the cache (and take up space, and CPU cycles on access) until the entire row is evicted. If a row remains in the cache indefinitely, so will the tombstones.

If the row is live, it's not a problem.

If a live row is accessed, then it deserves being in cache. Maybe it's occupying more space than it needs, but that's minor.

If a live row is not accessed, then it will be evicted eventually. Maybe it takes some time to get evicted, but that's minor.

The problem is frequently-accessed dead rows. They can accumulate, but do not contribute to output. Therefore the raio of live output bytes to cache bytes keeps getting worse. In queue style workloads (write new clustering rows, delete old clustering rows) this converts O(live rows) operations to O(live rows + dead rows).


> 
> AFAIK there is currently no plan to fix this. But if we had to fix it, I guess the solution would be to expire cell tombstones during version merges (in calls to `mutation_partition_v2::apply_monotonically` from cache) and during reads (in `cache_flat_mutation_reader::add_clustering_row_to_buffer`).
> 
> Does this deserve P1? Intuitively, it seems like a relatively minor optimization.

Some workloads are completely killed by the lack of this optimization. Queue workloads are advertised as an antipattern, but that just means people commonly apply it.

I don't think it's a P1. Expiring row tombstones probably addressed most of the workloads. Single-cell tombstones are probably an order of magnitude rarer, or perhaps even more.

**Comment by michoecho** [comment]
> If the row is live, it's not a problem.

Right, I thought the case of the dead row was already handled by https://github.com/scylladb/scylladb/pull/14463, but it only expires cells when it notices that the row marker is expired, and I forgot that UPDATE can update cell without updating the row marker.

**Comment by alezzqz** [comment]
> Here's another way, less artificial:
>     1. Have a schema with collections
>     2. add and remove cells from the collection
>     3. the collections in cache only grow, since tombstones aren't purged
> 
> 
> This one looks more real, but I can't tell whether it ever happens. I haven't seen cases where there is much delete activity on collections, but that doesn't mean a lot.

I'd suppose this case is quite real
My tests shows that when we add/remove single cell from collection SELECT from partition performance gets down 10 times(!) on just about 3000th iteration. It's obviously have to be fixed.
Unfortunately I have no clear idea how to fix it without impacting read performance yet
